Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE for linux/ResourceUsageThreadLinux

iidmsa opened the pull request:
#### fa8a900d329a5e2ef0eff49e553b081c30e32ca4
<pre>
Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE for linux/ResourceUsageThreadLinux
<a href="https://bugs.webkit.org/show_bug.cgi?id=294700">https://bugs.webkit.org/show_bug.cgi?id=294700</a>

Reviewed by Chris Dumez.

Replace unsafe C-Style fixed-size buffer with modern std::array and
std::span for safer buffer operations.

* Source/WebCore/page/linux/ResourceUsageThreadLinux.cpp:
(WebCore::threadCPUUsage):

Canonical link: <a href="https://commits.webkit.org/296489@main">https://commits.webkit.org/296489@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/02aad794a4bab1bd3d333ca9a12b6338726a6903

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108697 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28358 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18782 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113908 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59082 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110660 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29047 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36917 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82574 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111645 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23070 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97914 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63011 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22489 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16050 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58611 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92440 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16100 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117029 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35751 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26381 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91594 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36124 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94181 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91400 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23280 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36302 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14064 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31609 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35652 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41185 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35361 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38708 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37039 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/page/linux/ResourceUsageThreadLinux.cpp:
```diff
@@ -141,14 +141,12 @@ static bool threadCPUUsage(pid_t id, float period, ThreadInfo& info)
-    char buffer[BUFSIZ];
-    WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
-    buffer[0] = '\0';
+    std::array<char, BUFSIZ> buffer { };
```
Can we drop `{ }` and do `buffer[0] = '\0';` on the next line like the old code used to do? We don't need to zero out the whole array so changing the behavior here is a bit surprising.

cdumez commented on Source/WebCore/page/linux/ResourceUsageThreadLinux.cpp:
```diff
@@ -141,14 +141,12 @@ static bool threadCPUUsage(pid_t id, float period, ThreadInfo& info)
     ssize_t totalBytesRead = 0;
     while (totalBytesRead < maxBufferLength) {
-        ssize_t bytesRead = read(fd, buffer + totalBytesRead, maxBufferLength - totalBytesRead);
+        ssize_t bytesRead = read(fd, bufferSpan.data() + totalBytesRead, maxBufferLength - totalBytesRead);
```
`bufferSpan.subspan(totalBytesRead).data()` is safer.

cdumez commented on Source/WebCore/page/linux/ResourceUsageThreadLinux.cpp:
```diff
@@ -163,17 +161,21 @@ static bool threadCPUUsage(pid_t id, float period, ThreadInfo& info)
     close(fd);
-    buffer[totalBytesRead] = '\0';
+
+    if (static_cast<size_t>(totalBytesRead) < bufferSpan.size())
```
This change here is surprising. What was wrong with the old code?

iidmsa replied:
Wanted to add additional bounds checking also for extra safe buffer usage.

cdumez commented on Source/WebCore/page/linux/ResourceUsageThreadLinux.cpp:
```diff
@@ -163,17 +161,21 @@ static bool threadCPUUsage(pid_t id, float period, ThreadInfo& info)
     WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN // GLib port
     // FIXME: Use `find(std::span { buffer }, ')')` instead of `strchr()`.
-    char* position = strchr(buffer, ')');
+    char* position = strchr(bufferSpan.data(), ')');
```
Doesn't have to be in this PR but we have `WTF::find()` with spans as a safe replacement for `strchr()`.

cdumez commented on Source/WebCore/page/linux/ResourceUsageThreadLinux.cpp:
```diff
@@ -141,14 +141,12 @@ static bool threadCPUUsage(pid_t id, float period, ThreadInfo& info)
-    WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
-    buffer[0] = '\0';
+    std::array<char, BUFSIZ> buffer { };
+    auto bufferSpan = std::span(buffer).first(maxBufferLength);
```
Do we really need `.first(maxBufferLength)` here?

cdumez commented on Source/WebCore/page/linux/ResourceUsageThreadLinux.cpp:
```diff
@@ -141,14 +141,13 @@ static bool threadCPUUsage(pid_t id, float period, ThreadInfo& info)
-    WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
-    buffer[0] = '\0';
+    std::array<char, BUFSIZ> buffer;
+    auto bufferSpan = std::span(buffer);
```
We could just write:
```
std::span bufferSpan { buffer };
```

iidmsa committed:
Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE for linux/ResourceUsageThreadLinux
https://bugs.webkit.org/show_bug.cgi?id=294700

Reviewed by Chris Dumez.

Replace unsafe C-Style fixed-size buffer with modern std::array and
std::span for safer buffer operations.

* Source/WebCore/page/linux/ResourceUsageThreadLinux.cpp:
(WebCore::threadCPUUsage):

Canonical link: https://commits.webkit.org/296489@main

philn commented on Source/WebCore/page/linux/ResourceUsageThreadLinux.cpp:
```diff
@@ -168,12 +167,12 @@ static bool threadCPUUsage(pid_t id, float period, ThreadInfo& info)
     // Skip tid and name.
     WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN // GLib port
```
So can this be removed then?

cdumez replied:
I doubt it. The code below still uses `strchr()` and needs to be ported to `WTF::find()`.

He removed `WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN` for the other block above so some progress :)
